When data is sent to client, send http code after storage call

williamlardier opened the pull request:
- In case of error, we can then call the error.
- We do not need to destroy the response before we started to send data to the client: we just need to destroy the stream, if set, and return the right error to the client.
- No change when we started to receive bytes: we destroy the stream.
- When there is no error, we send the expected http header and potential headers before starting to send data.
- We centralize the logic to end the request and log.
- To better test, the http response mock is improved so we can really test streams.

Issue: ARSN-467

Seragonia committed:
When data is sent to client, send http code after storage call

- In case of error, we can then call the error.
- We do not need to destroy the response before we started to send
  data to the client: we just need to destroy the stream, if set,
  and return the right error to the client.
- No change when we started to receive bytes: we destroy the
  stream.
- When there is no error, we send the expected http header and
  potential headers before starting to send data.
- We centralize the logic to end the request and log.
- To better test, the http response mock is improved
  so we can really test streams.

Issue: ARSN-467

bert-e commented:
# Hello williamlardier,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://bert-e.scality.net/scality/arsenal/doc/user).


<details>
  <summary><b>Available options</b></summary>

  name   | description  | privileged   | authored
  ------ | ------------ | ------------ |----------
  `/after_pull_request` | Wait for the given pull request id to be merged before continuing with the current one. |  | 
  `/bypass_author_approval` | Bypass the pull request author's approval |  :star:  | 
  `/bypass_build_status` | Bypass the build and test status |  :star:  | 
  `/bypass_commit_size` | Bypass the check on the size of the changeset ```TBA``` |  :star:  | 
  `/bypass_incompatible_branch` | Bypass the check on the source branch prefix |  :star:  | 
  `/bypass_jira_check` | Bypass the Jira issue check |  :star:  | 
  `/bypass_peer_approval` | Bypass the pull request peers' approval |  :star:  | 
  `/bypass_leader_approval` | Bypass the pull request leaders' approval |  :star:  | 
  `/approve` | Instruct Bert-E that the author has approved the pull request. |  |  :writing_hand: 
  `/create_pull_requests` | Allow the creation of integration pull requests. |  | 
  `/create_integration_branches` | Allow the creation of integration branches. |  | 
  `/no_octopus` | Prevent Wall-E from doing any octopus merge and use multiple consecutive merge instead |  | 
  `/unanimity` | Change review acceptance criteria from `one reviewer at least` to `all reviewers` |  | 
  `/wait` | Instruct Bert-E not to run until further notice. |  | 
  

</details>



<details>
  <summary><b>Available commands</b></summary>

  name   | description  | privileged
  ------ | ------------ | ------------
  `/help` | Print Bert-E's manual in the pull request. | 
  `/status` | Print Bert-E's current status in the pull request ```TBA``` | 
  `/clear` | Remove all comments from Bert-E from the history ```TBA``` | 
  `/retry` | Re-start a fresh build ```TBA``` | 
  `/build` | Re-start a fresh build ```TBA``` | 
  `/force_reset` | Delete integration branches & pull requests, and restart merge process from the beginning. | 
  `/reset` | Try to remove integration branches unless there are commits on them which do not appear on the source branch. | 
  
</details>




*Status report is not available.*

bert-e commented:
# Incorrect fix version



The `Fix Version/s` in issue ARSN-467 contains:


* `8.1.148`

* `8.2.5`



Considering where you are trying to merge, I ignored possible hotfix versions and I expected to find:



* `8.2.5`


Please check the `Fix Version/s` of ARSN-467, or the target
branch of this pull request.

bert-e commented:
# Request integration branches



Waiting for integration branch creation to be requested by the user.

To request integration branches, please comment on this pull request with the following command:

```
/create_integration_branches
```

Alternatively, the `/approve` and `/create_pull_requests` commands will automatically
create the integration branches.

francoisferrand commented on lib/s3routes/routesUtils.ts:
```diff
@@ -441,23 +456,38 @@ function retrieveData(
+                // We write the http headers before piping the data to the
+                // response object.
+                // Any subsequent error leads to destror deletion.
+                if (!response.headersSent) {
```
this should always be `false`, right?

williamlardier replied:
Yes. As discussed this is more a security (lack of types where we do not have full confidence if we don't have a hidden route that sent the headers already). I can confirm that I see none currently. However, it will be handled in the long term ticket; ARSN-474 that will follow up this ticket with the long term rework.

francoisferrand commented on lib/s3routes/routesUtils.ts:
```diff
@@ -441,23 +456,38 @@ function retrieveData(
+                        // @ts-expect-error
+                        error: err.code, errMsg: err.message });
+                    // Only end if we haven't already destroyed the response
+                    if (!responseDestroyed) {
```
should we not just check `!response.headersSent` instead?

williamlardier replied:
Yes, code is updated.

francoisferrand commented on lib/s3routes/routesUtils.ts:
```diff
@@ -408,15 +420,18 @@ function retrieveData(
                         error: err,
                         method: 'retrieveData',
                     });
-                    _destroyResponse();
+                    _destroyReadable(readable);
```
we used to destroy the `response`, now we destroy the `request` : can't we just "not destroy anything", and just let the readable be destroyed normally? (the readable will be closed once the response is closed anyway)
or is there a reason/need for this explicit "early" destroy?

francoisferrand commented on lib/s3routes/routesUtils.ts:
```diff
@@ -317,9 +318,11 @@ function okContentHeadersResponse(
-    const httpCode = range ? 206 : 200;
-    log.debug('response http code', { httpCode });
-    response.writeHead(httpCode);
+    if (shouldWriteHeaders) {
```
instead of adding a condition here, I would rather remove `writeHead()` from here, and move it to the one place where `okContentHeadersResponse` is called with the default value shouldWriterHeaders=true (i.e. in `responseContentHeaders()`)

williamlardier replied:
Done.

francoisferrand commented on lib/s3routes/routesUtils.ts:
```diff
@@ -689,16 +718,15 @@ export function streamUserErrorPage(
-    return retrieveData(dataLocations, retrieveDataParams, response, log);
+    return retrieveData(dataLocations, retrieveDataParams, response, {
+        statusCode: error.code,
+        headers: { 'Content-type': 'text/html' },
```
nit: `'Content-type': 'text/html'` can be set *here* with `response.setHeader` ?
(and since it is a header, it can be "overwritten" later if we actually send an error?)

(so we don't need to pass the headers and complexify retrieveData)

williamlardier replied:
It was moved to a `setHeader`: it is basically the same at the end, but at least it simplifies how we pass the http code in the stream data function.

Seragonia committed:
Fix dependabot configuration targeting old branch

Now that the unified 8.2 exists, we remove the target branch
and directly use the default branch, as this is the documented
behavior of dependabot.

Issue: ARSN-467

Seragonia committed:
Add tests

ARSN-467

ghost commented:
/approve

Seragonia committed:
Test responseContentHeaders

Issue: ARSN-467

francoisferrand commented on lib/s3routes/routesUtils.ts:
```diff
@@ -441,23 +447,42 @@ function retrieveData(
-        },
+                currentStream = null;
+                if (err) {
+                    log.error('abort response due to error', {
```
`abort` is only in case the header was already sent, and we don't send an errorResponse. When we reach line 475, we are just sending a "regular" error.

williamlardier replied:
The logs were moved out of the `err` case, and consistently log once based on the error. Also we don't call `end` if the response was destroyed.

francoisferrand commented on lib/s3routes/routesUtils.ts:
```diff
@@ -441,23 +447,42 @@ function retrieveData(
+                        error: err.code, errMsg: err.message });
+                    if (!response.headersSent) {
+                        return XMLResponseBackend.errorResponse(errors.ServiceUnavailable, response, log);
+                    }
```
should we call `_destroyResponse` explicitely in the `else` case (to force aborting the response), or is it enough to call `response.end()` (i.e. no risk of closing the request cleanly but 'truncating' the body) ?

francoisferrand replied:
we _expect_ the response has been destroyed in other error case, but it is not clear or enforced: at least a comment should be added to explain  it.

Actually, I wonder if maybe we should just move this line, i.e.
```
- _destroyReadable(readable);
+ if (!response.headersSent) {
+     _destroyReadable(readable);
+     XMLResponseBackend.errorResponse(errors.ServiceUnavailable, response, log);
+ } else {
      _destroyResponse();
+ }
```

what do you think?

williamlardier replied:
The readable is for each location, hence we do not have it in the (error) callback, its handling should be implemented before, as might might have multiple locations...

francoisferrand replied:
(really not sure it is best : it makes the case more self-contained, but still requires to tweak error handling to skip the `end()`, and changes responses to be sent from multiple places...)

francoisferrand replied:
> The readable is for each location, hence we do not have it in the (error) callback, its handling should be implemented before, as might might have multiple locations...

yes, the "suggestion" (bad or good) was to send the response as well from within the loop (~line 425), were we close the readable

francoisferrand commented on lib/s3routes/routesUtils.ts:
```diff
@@ -441,23 +447,42 @@ function retrieveData(
+                // call end for all cases (error/success) per node.js docs
+                // recommendation
+                response.end(() => {
+                    log.end().info('responded with streamed content', {
```
when this happens after an error, the httpCode is not relevant anymore : the log does not match in that case...
It is also somewhat redundant with the error logs posted above.

williamlardier replied:
See first comment's answser

francoisferrand commented on lib/s3routes/routesUtils.ts:
```diff
@@ -408,15 +411,18 @@ function retrieveData(
+                    // In this case, we do not destroy the response:
+                    // nothing was sent to the client yet, so we
+                    // can return a proper error.
+                    log.debug('failed to get object', {
```
nit: not sure about the log level change. This "inner" log has all the context of the error, so may be more accurate... But at the same time indeed the final log is sure to be called, in every situation.

not sure which way to go, but if we change this error then the other logs must be consistent: i.e. `log.error('error piping data from source'` should also be changed to lower log level.

williamlardier replied:
I kept the error log, better dump twice than missing an error... And it remains consistent with other logs.

francoisferrand commented on tests/unit/s3routes/routesUtils/redirectRequestOnError.spec.js:
```diff
@@ -70,36 +70,26 @@ describe('routesUtils.redirectRequestOnError', () => {
-                mockedDataLocations[0]);
-            assert.strictEqual(dataWrapperGetStub.getCall(0).args[1],
-                responseMock);
+            dataWrapperGetStub.restore();
```
best to put this in a cleanup step, to ensure `restore()` even in case the test fails (so it does not affect other tests)

Seragonia committed:
Bump Arsenal version

Issue: ARSN-467

francoisferrand commented on lib/s3routes/routesUtils.ts:
```diff
@@ -441,23 +449,42 @@ function retrieveData(
+                // We write the http headers before piping the data to the
+                // response object.
+                // Any subsequent error leads to destror deletion.
+                // todo: need to check if headers were already set
```
not "todo", already done
```suggestion
```

williamlardier replied:
done

francoisferrand commented:
/create_integration_branches

bert-e commented:
# Conflict




A conflict has been raised during the creation of
integration branch `w/8.2/bugfix/ARSN-467-v2` with contents from `bugfix/ARSN-467-v2`
and `development/8.2`.

**I have not created the integration branch.**


Here are the steps to resolve this conflict:

```bash
 $ git fetch
 $ git checkout -B w/8.2/bugfix/ARSN-467-v2 origin/development/8.2
 $ git merge origin/bugfix/ARSN-467-v2
 $ # <intense conflict resolution>
 $ git commit
 $ git push -u origin w/8.2/bugfix/ARSN-467-v2
```





*The following options are set:* **approve**

bert-e commented:
# Build failed



The [build](https://github.com/scality/Arsenal/actions?query=branch%3Aw/8.2/bugfix/ARSN-467-v2) for [commit](https://github.com/scality/Arsenal/commit/9eb92f1b531611c51b29d8ebcf6835d194e3785a) did not succeed in branch [w/8.2/bugfix/ARSN-467-v2](https://github.com/scality/arsenal/compare/development/8.2...w/8.2/bugfix/ARSN-467-v2)



*The following options are set:* **approve**

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/8.1`

* :heavy_check_mark: `development/8.2`



The following branches have **NOT** changed:

* `development/6.4`
* `development/7.10`
* `development/7.4`
* `development/7.70`



Please check the status of the associated issue ARSN-467.

Goodbye williamlardier.




*The following options are set:* **approve**
